modules/core/packet-server/keeper/keeper.go

```diff
@@ -106,6 +107,112 @@ func (k Keeper) RecvPacket(
 	return nil
 }
 
+func (k Keeper) WriteAcknowledgement(
+	ctx sdk.Context,
+	_ *capabilitytypes.Capability,
+	packet exported.PacketI,
+	ack exported.Acknowledgement,
+) error {
+	// Lookup counterparty associated with our channel and ensure that it was packet was indeed
+	// sent by our counterparty.
+	counterparty, ok := k.ClientKeeper.GetCounterparty(ctx, packet.GetDestChannel())
+	if !ok {
+		return channeltypes.ErrChannelNotFound
+	}
+	if counterparty.ClientId != packet.GetSourceChannel() {
+		return channeltypes.ErrInvalidChannelIdentifier
+	}
+
+	// NOTE: IBC app modules might have written the acknowledgement synchronously on
+	// the OnRecvPacket callback so we need to check if the acknowledgement is already
+	// set on the store and return an error if so.
+	if k.ChannelKeeper.HasPacketAcknowledgement(ctx, packet.GetDestPort(), packet.GetDestChannel(), packet.GetSequence()) {
```

Comment:
We should be also checking that a receipt does exist. Otherwise, you can write an acknowledgement for a packet you never received

Reply:
It was not previously present. Remains to be seen with how it gets called with port-router-refactor, but generally i would expect this to be called by an app. Perhaps not directly but through the msg server